### PR TITLE
Fix 403 forbidden issue with '/api' and '/oapi' paths

### PR DESCRIPTION
**Pull request by nurali-techie**
Currently, oso-proxy drops '/api' and '/oapi' prefix in request path but OSIO UI (client) not sending '/api/api' and '/oapi/oapi' in Request path.

Fix the issue by supporting both '/api' and '/api/api' (same for oapi) in oso-proxy.

Fix - https://github.com/openshiftio/openshift.io/issues/4008

### CODE REVIEW

File: middlewares/osio/auth.go [hunk 1]
```diff
@@ -20,10 +20,12 @@ const (
...
-	metrics RequestType = "metrics"
-	console RequestType = "console"
-	logs    RequestType = "logs"
+	api      RequestType = "api"
+	oapi     RequestType = "oapi"
+	metrics  RequestType = "metrics"
+	console  RequestType = "console"
+	logs     RequestType = "logs"
+	undefine RequestType = ""
 )
```
Comment (nurali-techie) [inline middlewares/osio/auth.go]: `oapi` type added for '/oapi' paths.  now, rest all path will be referred as `undefine` type.

File: middlewares/osio/auth.go [hunk 2]
```diff
@@ -257,19 +261,36 @@ func (r RequestType) path() string {
...
+		pathPrefix = r.path() + r.path()
+		stripPath = r.path()
+	case metrics, console, logs:
+		pathPrefix = r.path()
+		stripPath = r.path()
+	case undefine:
+		return
+	}
+	stripRequestPathPrefix(req, pathPrefix, stripPath)
+}
```
Comment (nurali-techie) [inline middlewares/osio/auth.go]: strip path logic made more flexible.  depending on RequestType pathPrefix and stripPath value determine.  Now stripRequestPathPrefix() takes two args.

File: middlewares/osio/auth.go [hunk 3]
```diff
@@ -315,9 +336,9 @@ func cacheKey(plainKey string) string {
...
 
-func stripRequestPathPrefix(req *http.Request, prefix string) {
-	if strings.HasPrefix(req.URL.Path, prefix) {
-		req.URL.Path = stripPrefix(req.URL.Path, prefix)
+func stripRequestPathPrefix(req *http.Request, pathPrefix, stripPath string) {
+	if strings.HasPrefix(req.URL.Path, pathPrefix) {
+		req.URL.Path = stripPrefix(req.URL.Path, stripPath)
 		req.RequestURI = req.URL.RequestURI()
 	}
 }
```
Comment (nurali-techie) [inline middlewares/osio/auth.go]: now there are separate arg in stripRequestPathPrefix()
pathPrefix - what to match
stripPath - what to drop

File: middlewares/osio/auth_test.go [hunk 4]
```diff
@@ -326,3 +327,40 @@ func TestRemoveUserID(t *testing.T) {
...
+
+func TestStripPathPrefix(t *testing.T) {
+	tables := []struct {
+		reqType      RequestType
+		reqPath      string
+		expectedPath string
+	}{
+		{api, "/api/api/anything", "/api/anything"},
+		{api, "/api/anything", "/api/anything"},
+		{oapi, "/oapi/oapi/anything", "/oapi/anything"},
```
Comment (aslakknutsen) [inline middlewares/osio/auth_test.go]: I wouldn't consider oapi in it self a separate path, it belongs to api.

When client is configured to use the /api prefix you'd get /api/api and /api/oapi

**Pull request merged**